src/rx-query.ts

```diff
@@ -490,6 +486,121 @@ export class RxQueryBase<
+
+        // if this is a regular query, also load documents into cache
+        if (Array.isArray(value) && value.length > 0) {
+            const persistedQueryCacheIds = new Set(this._persistentQueryCacheResult);
+
+            let docsData: RxDocumentData<RxDocType>[] = [];
+
+            // query all docs updated > last persisted, limit to an arbitrary 1_000_000 (10x of what we consider our largest library)
+            const {documents: changedDocs} = await this.collection.storageInstance.getChangedDocumentsSince(
+              1_000_000,
+              // make sure we remove the monotonic clock (xxx.01, xxx.02) from the lwt timestamp to avoid issues with
+              // lookups in indices (dexie)
+              {id: '', lwt: Math.floor(Number(lwt))}
+            );
+
+            for (const changedDoc of changedDocs) {
+              /*
+               * no need to fetch again, we already got the doc from the list of changed docs, and therefore we filter
+               * deleted docs as well
+               */
+              persistedQueryCacheIds.delete(changedDoc[primaryPath] as string);
+
+              // ignore deleted docs or docs that do not match the query
+              if (!this.doesDocumentDataMatch(changedDoc)) {
+                continue;
+              }
+
+              // add to document cache
+              this.collection._docCache.getCachedRxDocument(changedDoc);
+
+              // add to docs
+              docsData.push(changedDoc);
+            }
+
+            // fetch remaining persisted doc ids
+            const nonRestoredDocIds: string[] = [];
+            for (const docId of persistedQueryCacheIds) {
+                // first try to fill from docCache
+                const docData = this.collection._docCache.getLatestDocumentDataIfExists(docId);
+                if (docData && this.doesDocumentDataMatch(docData)) {
+                  docsData.push(docData);
+                }
+
+                if (!docData) {
+                  nonRestoredDocIds.push(docId);
+                }
+            }
+
+            // otherwise get from storage
+            if (nonRestoredDocIds.length > 0) {
+              const docsMap = await this.collection.storageInstance.findDocumentsById(nonRestoredDocIds, false);
+              Object.values(docsMap).forEach(docData => {
+                this.collection._docCache.getCachedRxDocument(docData);
+                docsData.push(docData);
+              });
+            }
+
+            const normalizedMangoQuery = normalizeMangoQuery<RxDocType>(
+              this.collection.schema.jsonSchema,
+              this.mangoQuery
+            );
+            const sortComparator = getSortComparator(this.collection.schema.jsonSchema, normalizedMangoQuery);
+            const skip = normalizedMangoQuery.skip ? normalizedMangoQuery.skip : 0;
+            const limit = normalizedMangoQuery.limit ? normalizedMangoQuery.limit : Infinity;
```

Comment:
ugh @eliias i think that this approach actually has the same fundamental flaw the event-reduce library has with limit queries -- which we solve with the limit buffer.

imagine a world where you have 10000 docs match, but LIMIT 10. then those first 10 docs are deleted/changed (or even one of them) so that they no longer match the query. the correct thing to do here would be to load the next 10 docs that match the query. but you never cached those other 9990 that weren't in the limit. and those 9990 didn't change. so how do you get them?

I think you inevitably have to re-execute the whole query in this case. Otherwise your results will be incorrect. I think an easy way to check is just to make sure that your limit is equal to the length of your docsData. If it is, you just don't use the persisted query cache.

(By the way, this case isn't super common.)

(Also btw, persisting the limit buffer should let us mostly fix this...)

Reply:
Sorry I didn't think of this earlier :/

Reply:
copying my quick thoughts from slack here for visibility:

> I have instrumented the cost of a “write”, and it takes about 40-50ms. It is quite slow in comparison to when we’d write directly to IndexedDB (this is localforage), but it is not something a user would notice.

This is my main concern. You need to profile this on non-m1/m2 macbooks, as well as iphone and (especially) android devices

> Based on our other conversations on this PR I expect this to be a somewhat controversial discussion (sync writes), but I prefer this. Storing the limit buffer itself is just moving the problem elsewhere, because we then need to persist the limit buffer every time the result set changes as well.
This is because a user could also indirectly move an item that is in the limit buffer (aka the next skip: 50 query) from inbox -> later. When we then restore a stale limit buffer (the same way I restore the stale query cache in this PR) it would be returning wrong results as well.

limit buffer has a mechansim to handle this already. it simply re-execs the full query if something in the buffer has changed. it’s not perfect, but i think in practice in our app 95%+ of the time stuff won’t re-exec. so yes this is true:

> If we instead bust the query cache every time we discover a mismatch (different counts), we essentially trade write latency with query latency. I think this on its own would be fine, especially because we mostly restore the query cache to mitigate the cold-start issue before the memory storage initialization is done. Without continuously persisting the query cache, we loose this benefit on the next reload after a change. So the worst case here is that a user has a slow startup time on every other tab/window.

but you’re trading N writes for N queries every time a relevant doc updates, for a re-exec 5% (or less) of the time. Who is this for?